Mov-17,18,19,20 + pull genre with eventListner

Sergey113222 opened the pull request:
MOV-17,18,19,20

Sergey113222 committed:
1. create MovieController
2. create MovieConverter
3. create MovieService

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java:
```diff
@@ -0,0 +1,53 @@
+
+    @Override
+    public Long addMovie(MovieDto movieDto) {
+        return movieRepository.save(movieConverter.toModel(movieDto)).getId();
```
add check on existing movie before trying to persist it

RomanKashenok replied:
Query(select m from Movie m where m.externalId = :externalId") should be added to MovieRepository

Sergey113222 replied:
done

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java:
```diff
@@ -0,0 +1,53 @@
+    private final MovieConverter movieConverter;
+
+    @Override
+    public Long addMovie(MovieDto movieDto) {
```
addToFavouriteMovies

Sergey113222 replied:
rename all methods done

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java:
```diff
@@ -0,0 +1,53 @@
+@Service
+@AllArgsConstructor
+public class MovieServiceImpl implements MovieService {
+
```
User u = new User() - stubbed user

Sergey113222 committed:
fix add method MovieServiceImpl (if movie exist return existed id, else save movie)
add external_id  to movie and MovieDto
add to MovieRepository additional method findByExternalId

Sergey113222 committed:
fix that you can add new user (add logik with profile) becouse without profile I can not save or update user

add to postman write json request for save and update body

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/converter/impl/ProfileConverterImpl.java:
```diff
@@ -0,0 +1,37 @@
+import info.movietrash.cinemabase.dto.ProfileDto;
+import info.movietrash.cinemabase.model.Profile;
+
+public class ProfileConverterImpl implements ProfileConverter {
```
make it spring component

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/converter/impl/UserConverterImpl.java:
```diff
@@ -14,17 +14,23 @@ public User toModel(UserDto userDto) {
             throw new IllegalArgumentException("Some of required fields is null: " + userDto);
         }
         User user = new User();
+        ProfileConverterImpl profileConverter = new ProfileConverterImpl();
```
inject it as component

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/converter/impl/UserConverterImpl.java:
```diff
@@ -14,17 +14,23 @@ public User toModel(UserDto userDto) {
     @Override
     public UserDto toDto(User user) {
         UserDto userDto = new UserDto();
+        ProfileConverterImpl profileConverter = new ProfileConverterImpl();
```
same here

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/dto/MovieDto.java:
```diff
@@ -26,5 +26,5 @@
     @JsonProperty("genre_ids")
     private List<Integer> genreIds;
-
+    private UserDto user;
```
че у тебя в фильме делают узеры?

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/service/impl/GenreServiceImpl.java:
```diff
@@ -0,0 +1,40 @@
+package info.movietrash.cinemabase.service.impl;
+
```
what is it?

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/service/impl/UserServiceImpl.java:
```diff
@@ -16,6 +16,7 @@
     private final UserRepository userRepository;
     private final UserConverter userConverter;
 
+    // @Transactional
```
why not transactinal?

Sergey113222 committed:
1.fix that you can add in pullRequest
2.fix method reloadData() for adding all genres from tmdb
3.fix GenreConverterImpl
4.add externalId for genre(table-model-dto)

Sergey113222 committed:
made method reloadData() for adding all genres from tmdb AUTOMATICALLY with help @EventListner

Sergey113222 committed:
update method reloadData() for adding all genres from tmdb AUTOMATICALLY with help @EventListner (solve problem with duplicate genre on every turn on app)

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/controller/GenreController.java:
```diff
@@ -0,0 +1,21 @@
+package info.movietrash.cinemabase.controller;
```
Remove it

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/service/impl/GenreServiceImpl.java:
```diff
@@ -0,0 +1,71 @@
+        objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+        JsonNode responseBody = objectMapper.readTree(genresJson);
+        JsonNode resultsMassive = responseBody.path(JSON_NODE_STR);
+        List<GenreDto> jsonGenreList = objectMapper.readValue(resultsMassive.toString(), new TypeReference<List<GenreDto>>() {
```
add catch clause

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/converter/impl/GenreConverterImpl.java:
```diff
@@ -0,0 +1,44 @@
+
+    @Override
+    public List<GenreDto> toDto(List<Genre> genreList) {
+        if (genreList != null) {
```
remove null check (or add CollectionUtils.isEmpty() )

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/converter/impl/GenreConverterImpl.java:
```diff
@@ -0,0 +1,44 @@
+public class GenreConverterImpl implements GenreConverter {
+    @Override
+    public List<Genre> toModel(List<GenreDto> genreDtoList) {
+        if (genreDtoList != null) {
```
the same

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/converter/impl/MovieConverterImpl.java:
```diff
@@ -0,0 +1,37 @@
+@Component
+public class MovieConverterImpl implements MovieConverter {
+    @Override
+    public Movie toModel(MovieDto movieDto) {
```
(@Nonnull MovieDto movieDto)

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/converter/impl/ProfileConverterImpl.java:
```diff
@@ -0,0 +1,39 @@
+@Component
+public class ProfileConverterImpl implements ProfileConverter {
+    @Override
+    public Profile toModel(ProfileDto profileDto) {
```
the samw

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/dto/GenreDto.java:
```diff
@@ -0,0 +1,11 @@
+import lombok.Data;
+
+@Data
+public class GenreDto {
```
ExternalGenreDto ?

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/service/GenreService.java:
```diff
@@ -0,0 +1,10 @@
+package info.movietrash.cinemabase.service;
```
remove it

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/service/impl/GenreServiceImpl.java:
```diff
@@ -0,0 +1,71 @@
+        JsonNode resultsMassive = responseBody.path(JSON_NODE_STR);
+        List<GenreDto> jsonGenreList = objectMapper.readValue(resultsMassive.toString(), new TypeReference<List<GenreDto>>() {
+        });
+        if (genreRepository.count() == 0) {
```
move it as first part of method

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/service/impl/GenreServiceImpl.java:
```diff
@@ -0,0 +1,71 @@
+
+    @Override
+    @EventListener(ApplicationReadyEvent.class)
+    public void reloadData() throws JsonProcessingException {
```
getGenresExternal

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/service/impl/GenreServiceImpl.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    @Override
+    @EventListener(ApplicationReadyEvent.class)
```
move EventListener to application start class

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/service/impl/GenreServiceImpl.java:
```diff
@@ -0,0 +1,71 @@
+
+    private final GenreRepository genreRepository;
+    private final RestTemplate restTemplate;
+    private final ObjectMapper objectMapper;
```
move to configuration and make it @Bean or initialize as final field

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/service/impl/GenreServiceImpl.java:
```diff
@@ -0,0 +1,71 @@
+import java.net.URI;
+import java.util.List;
+
+@Service
```
@RequiredArgsConstructor

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java:
```diff
@@ -0,0 +1,68 @@
+        }
+        Movie movie = movieConverter.toModel(movieDto);
+        movie.getUsers().add(user.get());
+        return movieRepository.save(movie).getId();
```
rewrite to save user_movie

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java:
```diff
@@ -0,0 +1,68 @@
+
+    @Transactional
+    @Override
+    public Long addToFavouriteMovies(MovieDto movieDto) {
```
fixe like on video

RomanKashenok commented:
Fix conflicts! (merge branche 'development' into current)

Sergey113222 committed:
1. fix @many-to-many (user_movie) relationship
2. fix method getGenresExternal() with @EventListener and get start it in CinemaBaseApplication
3. fix UserMovie (add userId and MovieId)
4. fix method addToFavouriteMovies() in MovieServiceImpl

Sergey113222 committed:
1. fix @many-to-many (user_movie) relationship
2. fix method getGenresExternal() with @EventListener and get start it in CinemaBaseApplication
3. fix UserMovie (add userId and MovieId)
4. fix method addToFavouriteMovies() in MovieServiceImpl
5. fix pom.xml conflict

Sergey113222 committed:
Merge branch 'development' into Mov-17,18,19,20

# Conflicts:
#	pom.xml
#	src/main/java/info/movietrash/cinemabase/converter/impl/UserConverterImpl.java
#	src/main/java/info/movietrash/cinemabase/dto/MovieDto.java
#	src/main/java/info/movietrash/cinemabase/repository/GenreRepository.java
#	src/main/java/info/movietrash/cinemabase/repository/MovieRepository.java
#	src/main/resources/application.properties
#	src/main/resources/liquibase.properties

Sergey113222 committed:
1. fix @many-to-many (user_movie) relationship
2. fix method getGenresExternal() with @eventlistener and get start it in CinemaBaseApplication
3. fix UserMovie (add userId and MovieId)
4. fix method addToFavouriteMovies() in MovieServiceImpl
5. fix conflict

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/converter/impl/GenreConverterImpl.java:
```diff
@@ -11,34 +11,30 @@
+    public List<Genre> toModel(List<ExternalGenreDto> externalGenreDtoList) {
+
+        List<Genre> genreList = new ArrayList<>();
+        for (int i = 0; i < externalGenreDtoList.size(); i++) {
```
use Stream API here.
externalGenreDtoList.stream().map(....).collect(....)

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/converter/impl/GenreConverterImpl.java:
```diff
@@ -11,34 +11,30 @@
-            return genreDtoList;
+    public List<ExternalGenreDto> toDto(List<Genre> genreList) {
+
+        List<ExternalGenreDto> externalGenreDtoList = new ArrayList<>();
```
same as previous

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/exception/ErrorControllerAdvice.java:
```diff
@@ -0,0 +1,26 @@
+
+@ControllerAdvice
+@Slf4j
+public class ErrorControllerAdvice {
```
Rename it to AooExceptionHandler or something like that

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java:
```diff
@@ -19,21 +21,35 @@
-        if (movieRepository.findByExternalId(movieDto.getExternalId()) != null) {
-            return movieRepository.findByExternalId(movieDto.getExternalId()).getId();
+        Long userId = 1L;     // in my db have created user with id = 1
+        User user = null;
```
unnecessary variable

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java:
```diff
@@ -19,21 +21,35 @@
-            return movieRepository.findByExternalId(movieDto.getExternalId()).getId();
+        Long userId = 1L;     // in my db have created user with id = 1
+        User user = null;
+        Movie movie = null;
```
the same

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java:
```diff
@@ -19,21 +21,35 @@
+            user = userOptional.get();
+        }
+        Optional<Movie> movieOptional = Optional.ofNullable(movieRepository.findByExternalId(movieDto.getExternalMovieId()));
+        if (movieOptional.isPresent()) {
```
replace if-else with
Movie movie = movieOptional.getOrDefault(...)

RomanKashenok replied:
а почему все же не getOrDefault ?

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/service/impl/SearchServiceImpl.java:
```diff
@@ -0,0 +1,105 @@
+                .path("/3/search/movie")
+                .queryParam("api_key", apiKey)
+                .queryParam("query", dto.getQuery())
+                .queryParam("language", dto.getLang())
```
searchDto has only nullable field, so query and lang could be null

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/service/impl/SearchServiceImpl.java:
```diff
@@ -0,0 +1,105 @@
+        URI uri = UriComponentsBuilder.newInstance()
+                .scheme(scheme)
+                .host(host)
+                .path("/3/movie/latest")
```
move path to .properties file

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/service/impl/SearchServiceImpl.java:
```diff
@@ -0,0 +1,105 @@
+        URI uri = UriComponentsBuilder.newInstance()
+                .scheme(scheme)
+                .host(host)
+                .path("/3/movie/popular")
```
move path to .properties file

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/service/impl/SearchServiceImpl.java:
```diff
@@ -0,0 +1,105 @@
+        URI uri = UriComponentsBuilder.newInstance()
+                .scheme(scheme)
+                .host(host)
+                .path("/3/search/movie")
```
move path to .properties file

RomanKashenok reviewed (changes requested):
change it a little please :)

Sergey113222 committed:
09.06.2021 fix Get/Put/Delete/Post methods and tested them by Postman(save endpoints) - all ok

Sergey113222 committed:
10.06.2021 add ResponseEntity to Get/Put/Delete/Post methods and made GlobalExceptionHandler @ControllerAdvice

Sergey113222 committed:
11.06.2021 add Validaton to Get/Put/Delete/Post methods UserMovieController

Sergey113222 committed:
14.06.2021 add test to Get/Put/Delete/Post methods MovieServiceImplTest

Sergey113222 committed:
15.06.2021 refactoring this branch

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/controller/UserMovieController.java:
```diff
@@ -23,8 +23,8 @@ public MovieDto findFavouriteMovieById(@PathVariable("id") Long id) {
     @PutMapping
-    public void updateFavouriteMovie(@RequestBody MovieDto movieDto) {
-        movieService.updateFavouriteMovie(movieDto);
+    public void updateFavouriteMovie(@RequestBody MovieDto movieDto, @RequestParam Long userMovieId) {
```
userMovieId - это зачем? Почему одно в теле, а второе в параметрах?
Может лучше отправлять userId в pathVariable, а в теле - только dto?

RomanKashenok commented on src/main/java/info/movietrash/cinemabase/service/MovieService.java:
```diff
@@ -9,7 +9,7 @@
     MovieDto fetchFavouriteMovieById(Long id);
 
-    void updateFavouriteMovie(MovieDto movieDto);
+    void updateFavouriteMovie(MovieDto movieDto, Long userMovieId);
```
подумай над Dto для класса UserMovie - не придется передавать два параметра :)

RomanKashenok reviewed (approved):
Я бы еще поработал над ним, но апрув поставлю.

Sergey113222 committed:
Merge remote-tracking branch 'origin/development' into Mov-17,18,19,20

# Conflicts:
#	src/main/resources/application.properties

Sergey113222 committed:
16.06.2021 fix conflicts
